### ISSUE TITLE
bash_execute_unix_command: cannot find keymap for command

### ISSUE THREAD

**Issue by twiecki** [ellie/atuin#962]
When I press the up-arrow-key I get `bash_execute_unix_command: cannot find keymap for command`. This is bash on OSX.

**Comment by conradludgate** [comment]
What version of bash are you running?

**Comment by twiecki** [comment]
I upgraded to `GNU bash, version 5.2.15(1)-release (aarch64-apple-darwin21.6.0)` now but same error.

**Comment by footlose** [comment]
I'm having the same issue. on   4.2.46(2)-release (x86_64-redhat-linux-gnu)

**Comment by shenshouer** [comment]
the same issue on centos 7 Linux xxx 4.14.141-2.el7.x86_64 #1 SMP Tue Feb 15 15:56:56 CST 2022 x86_64 x86_64 x86_64 GNU/Linux

**Comment by orsenthil** [comment]
I notice this issue on Mac, when with the latest version of atuin

```
bash --version
GNU bash, version 5.1.16(1)-release (x86_64-apple-darwin20.6.0)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

> -bash: bash_execute_unix_command: cannot find keymap for command

**Comment by njames** [comment]
I am also having this issue - I just reinstalled atuin and restarted the terminal

```
bash --version
GNU bash, version 5.2.15(1)-release (x86_64-apple-darwin22.1.0)
```

**Comment by njames** [comment]
I reread the documentation and made these adjustments
```
 export ATUIN_NOBIND="true"
 eval "$(atuin init bash)"
 
 # bind to ctrl-r, add any other bindings you want here too
  bind -x '"\C-r": __atuin_history'
  bind  '"\e[A":history-search-backward'
  bind  '"\e[B":history-search-forward'
```

removing the -x seemed to help but enter is not returning the selected command from the  ctrl-r screen :(

**Comment by Subfire** [comment]
same issue on CentOS 7

**Comment by arcuru** [comment]
Resh has this issue as well and also hasn't resolved it: https://github.com/curusarn/resh/issues/157

**Comment by berceanu** [comment]
I have the same problem, with 
```
GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)
Copyright (C) 2011 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```
and 
`atuin 17.0.1`

**Comment by ryanwwest** [comment]
Having the same problem on macos with `GNU bash, version 3.2.57(1)-release (arm64-apple-darwin21)`. I installed this and it worked well for a few days until maybe a reboot?, and since then just this error.

**Comment by turian** [comment]
I just installed atuin on OSX and got this too.

GNU bash, version 3.2.57(1)-release (arm64-apple-darwin23)

**Comment by akinomyoga** [comment]
This was just fixed in commit https://github.com/atuinsh/atuin/pull/1515/commits/b18f7acc52f57d557688f23a1c5036406e0525e5 in #1515 two days ago. It will be fixed in the next release. Anyway, you can instead install the latest version of Bash. The problem doesn't happen with `bash >= 4.3`. The 17-year-old Bash 3.2 has other issues, so it is generally recommended to install the latest version of Bash.

**Comment by berceanu** [comment]
That is not always possible, for example on HPC systems where the user does not have administrative privileges.

**Comment by akinomyoga** [comment]
Of course, it is not always possible. It's just *generally recommended*. I meant "usually" by the term "generally", and the recommendation is just a recommendation. Anyway, I think it's still worth considering.

If you are on an HPC system, you *usually* (*not 100%*) have more chances than on the other types of systems. We usually have compilers in HPC systems, so you can compile Bash from the source code and install it in your home directory. You can `exec` your local Bash in `.bashrc` with a version check on the currently running Bash. Sometimes, compilers on HPC systems produce strange binaries whose forking and threading ability is restricted so that the job scheduler can interfere with it. Even in such a case, you may try cross-compiling (with a matching library) in another host or generate a standalone binary with `-static*` options, and copy the generated binary to the HPC system. Or you should ask the administrator if it is possible to introduce a newer version of Bash (probably after some approval by some committee meeting, the head of the institution, etc.). It's still not 100% percent, but it's worth trying.

**Comment by MaxenceG2M** [comment]
Recently updated my atuin to version 18.10.0, I have the same error with `Ctrl+G` shortcut.

```sh
$ bash --version
GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)
[...]
$ atuin --version
atuin 18.10.0 (37e257105ab801b49fd07a927d707a36a2e771a1)
```

No issues with atuin 18.9.0

**Comment by akinomyoga** [comment]
@MaxenceG2M Indeed, we have changed keybindings for Bash in 18.10.0, but I cannot reproduce the reported behavior (assuming that you meant that you see the error message `bash_execute_unix_command: cannot find keymap for command`).

What is the result of `echo "$BASH_VERSION"`?

As far as I know, the error message `bash_execute_unix_command: cannot find keymap for command` was caused by an internal bug of Bash, and it is already fixed in recent versions of Bash >= 5.0. If it still happens in 5.1, it is a bug in Bash 5.1. However, on this thread, there are too many people reporting a wrong Bash version obtained by `bash --version` (in particular from macOS). The command `bash --version` prints the version of *Bash that is currently found in `PATH`*. It doesn't necessarily match the version of *Bash that is started by the terminal emulator* (which is determined based on the `passwd` entry or the environment variable `SHELL` or the terminal configuration, depending on the terminal). The correct way to check the version of Bash that is currently running is to check the shell variable `$BASH_VERSION`.

Anyway, I tried Bash 3.2, 4.2, 4.3, 4.4, 5.0, and 5.1, but I couldn't reproduce the reported behavior in any versions of Bash. This implies that there are some specific configurations that cause the problem, or that there is another factor. @MaxenceG2M If you have an idea, could you explain your detailed setup to reproduce the problem?

**Comment by MaxenceG2M** [comment]
Thank a lot for you answer !

**TL;DR**: I have a kind of conflict with [fzf](https://github.com/junegunn/fzf)...

---

First of all, the copy/paste of the exact error message when I press `Ctrl+G`:
```sh
bash: bash_execute_unix_command: cannot find keymap for command
```

The `BASH_VERSION` var (sadly the same as `bash --version`):
```sh
$ echo $BASH_VERSION
5.1.16(1)-release
```

I absolutely don't understand why I have this problem... So I remove everything in my `bashrc` to understand was going on.

My cleaned and very little `bashrc` file:
```bashrc
[ -f ~/.fzf.bash ] && source ~/.fzf.bash # <-- removing this line solve everything
function fzf-repo() {
    echo hello
}
bind -x '"\C-g": fzf-repo'

function fzf-kubectx() {
    echo coucou 
}
bind -x '"\C-k": fzf-kubectx'

. "$HOME/.atuin/bin/env"
[[ -f ~/.bash-preexec.sh ]] && source ~/.bash-preexec.sh
eval "$(atuin init bash --disable-up-arrow)"
```

So there is a kind of conflict between atuin and fzf. Why, that is the question :D

**Comment by akinomyoga** [comment]
Thank you for identifying the setup. I'll try this later. If you could share your `fzf` version, it would help (as the detailed implementation of `fzf`'s Bash integration also depends on its version).

**Comment by akinomyoga** [comment]
@MaxenceG2M I could reproduce the error. The minimum reproducer of the problem is this:

```bash
# bashrc

bind -x '"\C-r": echo C-r'
bind -x '"\C-g": echo C-g'
bind '"\C-r": ""'
```

With this `~/.bashrc`, pressing <kbd>C-g</kbd> prints the error message `bash_execute_unix_command: cannot find keymap for command`. The problem reproduces in Bash 5.1, but not in 5.0 and 5.2. I searched the mailing list of Bash and found [the corresponding report](https://lists.gnu.org/archive/html/bug-bash/2021-04/msg00135.html), which was fixed soon. I actually [submitted a patch](https://lists.gnu.org/archive/html/bug-bash/2021-04/msg00136.html) then, though Chet, the Bash maintainer, seems to have also fixed the bug in parallel.

You can update your Bash to the latest version, or, as a workaround, it seems you can unbind `\C-r` explicitly before initializing Atuin.

```diff
 [ -f ~/.fzf.bash ] && source ~/.fzf.bash # <-- removing this line solve everything
 function fzf-repo() {
     echo hello
 }
 bind -x '"\C-g": fzf-repo'

 function fzf-kubectx() {
     echo coucou 
 }
 bind -x '"\C-k": fzf-kubectx'

 . "$HOME/.atuin/bin/env"
 [[ -f ~/.bash-preexec.sh ]] && source ~/.bash-preexec.sh
+bind -r '\C-r'
 eval "$(atuin init bash --disable-up-arrow)"
```

**Comment by akinomyoga** [comment]
I'll later submit a PR for the workaround in Atuin's shell integration.

**Comment by akinomyoga** [comment]
> I'll later submit a PR for the workaround in Atuin's shell integration.

#2975

**Comment by ellie** [comment]
Resolved by #2975
